docs/codeql/writing-codeql-queries/debugging-data-flow-queries-using-partial-flow.rst

```diff
@@ -0,0 +1,116 @@
+.. _debugging-data-flow-queries-using-partial-flow:
+
+Debugging data-flow queries using partial flow
+==============================================
+
+If a data-flow query doesn't produce the results you expect to see, you can use partial flow to debug the problem.
+
+In CodeQL, you can use :ref:`data flow analysis <about-data-flow-analysis>` to compute the possible values that a variable can hold at various points in a program.
+A typical data-flow query looks like this:
+
+.. code-block:: ql
+
+::
+
+    class MyConfig extends TaintTracking::Configuration {
+      MyConfig() { this = "MyConfig" }
+
+      override predicate isSource(DataFlow::Node node) { node instanceof MySource }
+
+      override predicate isSink(DataFlow::Node node) { node instanceof MySink }
+    }
+
+    from MyConfig config, DataFlow::PathNode source, DataFlow::PathNode sink
+    where config.hasFlowPath(source, sink)
+    select sink.getNode(), source, sink, "Sink is reached from $@.", source.getNode(), "here"
+
+The same query can be slightly simplified by rewriting it without :ref:`path explanations <creating-path-queries>`:
+
+.. code-block:: ql
+
+    from MyConfig config, DataFlow::Node source, DataFlow::Node sink
+    where config.hasPath(source, sink)
+    select sink, "Sink is reached from $@.", source.getNode(), "here"
+
+If a data-flow query that you have written doesn't produce the results you expect it to, there may be a problem with your query.
+You can try to debug the potential problem by following the steps described below. 
+
+Checking sources and sinks
+--------------------------
+
+Initially, you should make sure that the ``source`` and ``sink`` definitions contain what you expect. If either the ``source`` or ``sink`` is empty then there can never be any data flow. The easiest way to check this is using quick evaluation in CodeQL for VS Code. Select the text ``node instanceof MySource``, right-click, and choose "CodeQL: Quick Evaluation". This will evaluate the highlighted text, which in this case means the set of sources. For more information, see :ref:`Analyzing your projects <running-a-specific-part-of-a-query-or-library>` in the CodeQL for VS Code help.
```

Comment:
The words `source` and `sink` are not referring directly to code, so I don't think they should be type-set specially.

Reply:
This was changed throughout, and I think it should be reverted throughout.